### PR TITLE
Feature 10 : All PUT APIs

### PR DESCRIPTION
**Pull request by Abhishek-Jain-1925**
Feature 10 : All PUT APIs

Work Done are as follows - 
1. Handling Request of All PUT APIs
2. Handling Request of List Skills Static API
3. Handler, Service and Repo layered Working Setup
4. Changed Prev. code as per suggestions
5. Unit Test Cases of Each Module with 92% coverage.
6. Applied Suggested changes as per Demo of API meet-up.

### CODE REVIEW

File: internal/api/handler/achievement_handler.go [hunk 1]
```diff
@@ -40,3 +48,41 @@ func CreateAchievementHandler(ctx context.Context, profileSvc service.Service) f
...
+		}
+
+		err = req.Validate()
+		if err != nil {
+			middleware.ErrorResponse(w, http.StatusBadRequest, err)
+			zap.S().Error(err)
+			return
+		}
+
+		value, err := achSvc.UpdateAchievement(ctx, profileID, eduID, req)
```
Comment (OmkeshJosh) [inline internal/api/handler/achievement_handler.go]: updatedResp, err := achSvc.UpdateAchievement(ctx, profileID, eduID, req)

File: internal/api/handler/achievement_handler.go [hunk 2]
```diff
@@ -40,3 +48,41 @@ func CreateAchievementHandler(ctx context.Context, profileSvc service.Service) f
...
+		if err != nil {
+			middleware.ErrorResponse(w, http.StatusBadGateway, err)
+			zap.S().Error("Unable to update achievement : ", err, "for profile id : ", profileID)
+			return
+		}
+
+		middleware.SuccessResponse(w, http.StatusOK, dto.MessageResponseWithID{
+			Message:   "Achievement updated successfully",
+			ProfileID: value,
+		})
```
Comment (OmkeshJosh) [inline internal/api/handler/achievement_handler.go:120]: Add Success logs in update,delete,create orations 

```
Create - > profile created with profile id
Update -> profile update with profile id, req body (if not sensitive info)
Delete  -> profile deleted with profile id

```


It help while debuging

File: internal/api/tests/profile_handler_test.go [hunk 3]
```diff
@@ -185,23 +186,70 @@ func TestGetProfileListHandler(t *testing.T) {
...
+
+	tests := []struct {
+		name               string
+		setup              func(mock *mocks.Service)
+		expectedStatusCode int
+	}{
+		{
+			name: "Success for listing skills",
+			setup: func(mockSvc *mocks.Service) {
+				mockListSkills := dto.ListSkills{Name: []string{"GO", "RUBY", "C", "C++", "JAVA", "PYTHON", "JAVASCRIPT"}}
```
Comment (OmkeshJosh) [inline internal/api/tests/profile_handler_test.go]: use constant

File: internal/app/service/achievement_service.go [hunk 4]
```diff
@@ -41,5 +50,27 @@ func (achSvc *service) CreateAchievement(ctx context.Context, cDetail dto.Create
...
+	if err != nil {
+		zap.S().Error("error to get achievement params : ", err, " for profile id : ", profileID)
+		return 0, err
+	}
+
+	var value repository.UpdateAchievementRepo
+	value.Name = req.Achievement.Name
+	value.Description = req.Achievement.Description
+	value.UpdatedAt = today
+	value.UpdatedByID = 1
```
Comment (OmkeshJosh) [inline internal/app/service/achievement_service.go:65]: add comment why we are using static value or add TODO comment

File: internal/repository/project_repository.go [hunk 5]
```diff
@@ -99,3 +100,30 @@ func (projectStore *ProjectStore) GetProjects(ctx context.Context, profileID int
 
 	return values, nil
 }
+
+// UpdateProject updates projects details into the database.
+func (projectStore *ProjectStore) UpdateProject(ctx context.Context, profileID int, eduID int, req UpdateProjectRepo) (int, error) {
+	updateQuery, args, err := sq.Update("projects").Set("name", req.Name).
```
Comment (OmkeshJosh) [inline internal/repository/project_repository.go]: use SetMap()

File: internal/repository/profile_repository.go [hunk 6]
```diff
@@ -77,13 +78,24 @@ func (profileStore *ProfileStore) ListProfiles(ctx context.Context) (values []dt
...
 		var value dto.ListProfiles
-		rows.Scan(&value.ID, &value.Name, &value.Email, &value.YearsOfExperience, &value.PrimarySkills, &value.IsCurrentEmployee)
+		var isCurrentEmployee int
+		err := rows.Scan(&value.ID, &value.Name, &value.Email, &value.YearsOfExperience, &value.PrimarySkills, &isCurrentEmployee)
+		if err != nil {
+			zap.S().Error("error scanning row:", err)
+			return []dto.ListProfiles{}, err
+		}
+
+		if isCurrentEmployee == 1 {
```
Comment (OmkeshJosh) [inline internal/repository/profile_repository.go]: handle it in service layer and use `map[bool]string{}{true:"Yes",false:"No"}`

try to keep independent db level no any service layer code, later on may be we need to change repo type pgSql -> mySql

File: internal/repository/experience_repository.go [hunk 7]
```diff
@@ -96,3 +97,31 @@ func (expStore *ExperienceStore) GetExperiences(ctx context.Context, profileID i
 
 	return values, nil
 }
+
+// UpdateExperience updates experience details into the database.
+func (expStore *ExperienceStore) UpdateExperience(ctx context.Context, profileID int, eduID int, req UpdateExperienceRepo) (int, error) {
+	updateQuery, args, err := sq.Update("experiences").Set("designation", req.Designation).
```
Comment (OmkeshJosh) [inline internal/repository/experience_repository.go]: use SetMap

File: internal/pkg/dto/project.go [hunk 8]
```diff
@@ -94,3 +96,36 @@ func (p *Project) Validate() error {
 	}
 	return nil
 }
+
+// Validate func checks if the Project details are valid.
+func (p *UpdateProjectRequest) Validate() error {
+
+	if p.Project.Name == "" {
```
Comment (OmkeshJosh) [inline internal/pkg/dto/project.go]:
``` 
fields := map[string]string{
        "name":                 p.Project.Name,
        "description":          p.Project.Description,
        "role":                 p.Project.Role,
        "responsibilities":     p.Project.Responsibilities,
        "technologies":         p.Project.Technologies,
        "technology worked on": p.Project.TechWorkedOn,
        "working start date":   p.Project.WorkingStartDate,
        "working end date":     p.Project.WorkingEndDate,
        "duration":             p.Project.Duration,
    }

    for fieldName, fieldValue := range fields {
        if fieldValue == "" {
            return fmt.Errorf("%s : %s", errors.ErrParameterMissing.Error(), fieldName)
        }
    }
```

File: internal/api/handler/achievement_handler.go (hunk 1 above)
Comment (Abhishek-Jain-1925) [reply internal/api/handler/achievement_handler.go]: Done!

File: internal/api/handler/achievement_handler.go (hunk 2 above)
Comment (Abhishek-Jain-1925) [reply internal/api/handler/achievement_handler.go:120]: Done!

File: internal/api/tests/profile_handler_test.go (hunk 3 above)
Comment (Abhishek-Jain-1925) [reply internal/api/tests/profile_handler_test.go]: Done!

File: internal/repository/project_repository.go (hunk 5 above)
Comment (Abhishek-Jain-1925) [reply internal/repository/project_repository.go]: Done!

File: internal/repository/profile_repository.go (hunk 6 above)
Comment (Abhishek-Jain-1925) [reply internal/repository/profile_repository.go]: Done!

File: internal/pkg/dto/project.go (hunk 8 above)
Comment (Abhishek-Jain-1925) [reply internal/pkg/dto/project.go]: Done!

File: internal/repository/experience_repository.go (hunk 7 above)
Comment (Abhishek-Jain-1925) [reply internal/repository/experience_repository.go]: Done!

File: internal/api/handler/achievement_handler.go [hunk 9]
```diff
@@ -27,16 +35,54 @@ func CreateAchievementHandler(ctx context.Context, profileSvc service.Service) f
...
 			Message:   "Achievement(s) added successfully",
 			ProfileID: profileID,
 		})
 	}
 }
+
+// UpdateAchievementHandler returns an HTTP handler that updates achievement using profileSvc.
+func UpdateAchievementHandler(ctx context.Context, achSvc service.Service) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		profileID, eduID, err := helpers.GetMultipleParams(r)
```
Comment (OmkeshJosh) [inline internal/api/handler/achievement_handler.go:56]: Create common function like 
- GetStringPathParam(r, constants.ProfileIDStr) // it return string type val
- GetIntPathParam((r, constants.EducationIDStr) // it return int type val

and define path param in constant pkg eg - profile_id, id

File: internal/api/handler/achievement_handler.go [hunk 10]
```diff
@@ -27,16 +35,54 @@ func CreateAchievementHandler(ctx context.Context, profileSvc service.Service) f
...
+		if err != nil {
+			middleware.ErrorResponse(w, http.StatusBadRequest, err)
+			zap.S().Error(err)
+			return
+		}
+
+		updatedResp, err := achSvc.UpdateAchievement(ctx, profileID, eduID, req)
+		if err != nil {
+			middleware.ErrorResponse(w, http.StatusBadGateway, err)
+			zap.S().Error("Unable to update achievement : ", err, "for profile id : ", profileID)
```
Comment (OmkeshJosh) [inline internal/api/handler/achievement_handler.go:80]: add edu id in logs

File: internal/api/handler/achievement_handler.go [hunk 11]
```diff
@@ -27,16 +35,54 @@ func CreateAchievementHandler(ctx context.Context, profileSvc service.Service) f
...
+			return
+		}
+
+		middleware.SuccessResponse(w, http.StatusOK, specs.MessageResponseWithID{
+			Message:   "Achievement updated successfully",
```
Comment (OmkeshJosh) [inline internal/api/handler/achievement_handler.go:85]: Show proper msg with eduID and ProfileID

File: internal/api/handler/certificate_handler.go [hunk 12]
```diff
@@ -5,14 +5,22 @@ import (
...
 	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/middleware"
+	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/specs"
 	"go.uber.org/zap"
 )
 
 // CreateCertificateHandler handles HTTP requests to add certificates details to a user profile.
-func CreateCertificateHandler(ctx context.Context, profileSvc service.Service) func(http.ResponseWriter, *http.Request) {
+func CreateCertificateHandler(ctx context.Context, certificateSvc service.Service) func(http.ResponseWriter, *http.Request) {
 	return func(w http.ResponseWriter, r *http.Request) {
+		id, err := helpers.GetParamsByID(r, "profile_id")
```
Comment (OmkeshJosh) [inline internal/api/handler/certificate_handler.go:18]: use this fun in line internal/api/handler/achievement_handler.go:55 

also define new helper function for string type.

File: internal/api/handler/decoder.go [hunk 13]
```diff
@@ -4,93 +4,139 @@ import (
...
 
 // Decodes the Profile Project object request
-func decodeCreateProjectRequest(r *http.Request) (dto.CreateProjectRequest, error) {
-	var req dto.CreateProjectRequest
+func decodeCreateProjectRequest(r *http.Request) (specs.CreateProjectRequest, error) {
+	var req specs.CreateProjectRequest
 	err := json.NewDecoder(r.Body).Decode(&req)
 	if err != nil {
-		return dto.CreateProjectRequest{}, errors.ErrInvalidBody
+		return specs.CreateProjectRequest{}, errors.ErrInvalidBody
```
Comment (OmkeshJosh) [inline internal/api/handler/decoder.go:73]: log the error

File: internal/app/service/achievement_service.go [hunk 14]
```diff
@@ -3,32 +3,36 @@ package service
...
+	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/specs"
 	"github.com/joshsoftware/profile_builder_backend_go/internal/repository"
 	"go.uber.org/zap"
 )
 
 // AchievementService represents a set of methods for accessing the achievements
 type AchievementService interface {
-	CreateAchievement(ctx context.Context, cDetail dto.CreateAchievementRequest) (profileID int, err error)
+	CreateAchievement(ctx context.Context, cDetail specs.CreateAchievementRequest, ID int) (profileID int, err error)
+	UpdateAchievement(ctx context.Context, profileID string, eduID string, req specs.UpdateAchievementRequest) (id int, err error)
```
Comment (OmkeshJosh) [inline internal/app/service/achievement_service.go:16]: Keep same data type 
profileID int or profileID string

File: internal/app/service/achievement_service.go [hunk 15]
```diff
@@ -3,32 +3,36 @@ package service
...
-func (achSvc *service) CreateAchievement(ctx context.Context, cDetail dto.CreateAchievementRequest) (profileID int, err error) {
+func (achSvc *service) CreateAchievement(ctx context.Context, cDetail specs.CreateAchievementRequest, id int) (profileID int, err error) {
 	if len(cDetail.Achievements) == 0 {
 		zap.S().Error("achievements payload can't be empty")
 		return 0, errors.ErrEmptyPayload
 	}
-	var value []repository.AchievementDao
+
+	var value []repository.AchievementRepo
 	for i := 0; i < len(cDetail.Achievements); i++ {
```
Comment (OmkeshJosh) [inline internal/app/service/achievement_service.go:28]: use range

**Comment by OmkeshJosh** [comment]
Please make note or create new trello ticket for above points.

**Pull request merged**